metis-indexing/src/main/java/eu/europeana/indexing/FullBeanPublisher.java

```diff
@@ -159,4 +256,226 @@ private void publishToSolr(RdfWrapper rdf, FullBeanImpl fullBean) throws Indexin
       throw new RecordRelatedIndexingException(SOLR_SERVER_PUBLISH_ERROR, e);
     }
   }
+
+  private List<String> searchMatchingRecordForRedirection(RdfWrapper rdfWrapper,
+      List<String> datasetIdsToRedirectFrom)
+      throws IndexerRelatedIndexingException, RecordRelatedIndexingException {
+    //The incoming structure of the identifier is /datasetId/recordId
+    final String[] splitRecordIdentifier = rdfWrapper.getAbout().split("/");
+    String datasetId = splitRecordIdentifier[1];
+    String recordId = splitRecordIdentifier[2];
+
+    //Create combinations of all rules into one query
+    final String queryForDatasetIds = generateQueryForDatasetIds(datasetIdsToRedirectFrom,
+        recordId);
+    final Pair<String, Map<String, List<String>>> queryMatchingFieldsAndFirstQueryGroup = generateQueryForMatchingFields(
+        rdfWrapper);
+    final String queryForMatchingFields = queryMatchingFieldsAndFirstQueryGroup.getLeft();
+    final String combinedQueryOr = Stream.of(queryForDatasetIds, queryForMatchingFields)
+        .filter(StringUtils::isNotBlank).collect(Collectors.joining(" OR "));
+
+    //Create query to restrict search on specific datasetId subsets
+    final List<String> datasetIds = new ArrayList<>();
+    datasetIds.add(datasetId);
+    if (!CollectionUtils.isEmpty(datasetIdsToRedirectFrom)) {
+      datasetIdsToRedirectFrom.stream().filter(StringUtils::isNotBlank).forEach(datasetIds::add);
+    }
+    final String datasetIdSubsets = generateQueryInDatasetSubsets(datasetIds);
+    final String finalQuery = String.format("(%s) AND (%s)", datasetIdSubsets, combinedQueryOr);
+
+    //Apply solr query and execute
+    final Map<String, String> queryParamMap = new HashMap<>();
+    queryParamMap.put("q", finalQuery);
+
+    //Prepare sub-query
+    Predicate<SolrDocument> predicate = document -> {
+      boolean doesDocumentMatchWithQuery = false;
+      for (Entry<String, List<String>> entry : queryMatchingFieldsAndFirstQueryGroup.getRight()
+          .entrySet()) {
+        doesDocumentMatchWithQuery = document.getFieldValues(entry.getKey()).stream()
+            .map(String.class::cast)
+            .anyMatch(entry.getValue()::contains);
+      }
+      return doesDocumentMatchWithQuery;
+    };
+    //Preprocess sub-query and replace documents based on the result
+    SolrDocumentList solrDocuments = getSolrDocuments(queryParamMap);
+    final SolrDocumentList subQueryResults = solrDocuments.stream().filter(predicate)
+        .collect(Collectors.toCollection(SolrDocumentList::new));
+    if (!CollectionUtils.isEmpty(subQueryResults)) {
+      solrDocuments = subQueryResults;
+    }
+    //Return all identifiers found
+    if (!CollectionUtils.isEmpty(solrDocuments)) {
+      return solrDocuments.stream()
+          .map(document -> (String) document.getFieldValue(EdmLabel.EUROPEANA_ID.toString()))
+          .collect(Collectors.toList());
+    }
+    return new ArrayList<>();
```

Comment:
solrDocuments cannot be null at this point, so you can remove the if statement and return the result of the stream operation.

Reply:
Sure